### ISSUE TITLE
Posterior chained calls arguments being evaluated before the first calls

### ISSUE THREAD

**Issue by ghsoares** [godotengine/godot#78334]
### Godot version

v4.0.3.stable.official.5222a99f5

### System information

Windows 11

### Issue description

When using function call chaining from my custom class, it seems to evaluate posterior arguments before the chained calls. For example I script like the follow:
```gd
# test.gd
extends Node

class Test:
	func foo():
		print("foo")
		return self
	func bar(arg):
		print("bar: ", arg)
		return self
	func foobar():
		print("foobar")
		return 0.0
	func test1():
		print("test1")
		return 1.0

func _ready() -> void:
	var test = Test.new()
	test.foo().bar(test.foobar()).foo().bar(test.test1())
```
It's expected to it to print:
```
foo
foobar
bar: 0
foo
test1
bar: 1
```
But it seems to evaluate the arguments first right-to-left then the calls left-to-right:
```
test1
foobar
foo
bar: 0
foo
bar: 1
```
It can be a huge issue when the chained calls depends on the internal state of the previous calls, for example in my specific case I'm building a in-code animation utility simillar to Tween by chaining calls:
```gd
motion.prop("scale").key(motion.current()).ease_out(Vector2(1.25, 1.25), 0.1)
```
In this example, I need to set the property I want to animate first, then set the keyframes, `motion.current()` gets the current (current from node or from last added keyframe) value of the property track, in this case it would be the current node `scale` value, but it throws an error because it doesn't know from which property to get the current value, because it expects `prop` to be called first, then `current`, but it calls `current` first then `prop`

### Steps to reproduce

```gd
# test.gd
extends Node

class Test:
	func foo():
		print("foo")
		return self
	func bar(arg):
		print("bar: ", arg)
		return self
	func foobar():
		print("foobar")
		return 0.0
	func test1():
		print("test1")
		return 1.0

func _ready() -> void:
	var test = Test.new()
	test.foo().bar(test.foobar()).foo().bar(test.test1())
```
To compare the output without chaining:
```gd
class Test:
	func foo():
		print("foo")
		return self
	func bar(arg):
		print("bar: ", arg)
		return self
	func foobar():
		print("foobar")
		return 0.0
	func test1():
		print("test1")
		return 1.0

func _ready() -> void:
	var test = Test.new()

	test.foo()
	test.bar(test.foobar())
	test.foo()
	test.bar(test.test1())
```

### Minimal reproduction project

N/A

**Comment by AThousandShips** [comment]
> It's expected to it to print:
> ```
> foo
> foobar
> bar: 0
> foo
> test1
> bar: 1
> ```

This part is not really possible and doesn't match other languages such as C++

The reason for this is how things are computed, it's done as part of analysis and works from the end of the calls, however one could expect:
```
foobar
test1
foo
bar: 0
foo
bar: 1
```

It's not gonna be possible to delay the calling of the arguments until after the preceeding functions have been called I'd argue

**Comment by ghsoares** [comment]
> This part is not really possible and doesn't match other languages such as C++

I don't know if I get this part correctly, but when I tested a chain example of my library in C++ while printing for each function call:
```cpp
motion->prop("scale")->frame(motion->current())->ease_out(motion->relative(Vector2(0.25, 0.25)), 0.5);
```
The order of the calls printed is `prop`, `current`, `frame`, `relative` then `ease_out` as I've expected.

**Comment by AThousandShips** [comment]
My bad, got it confused with just the arguments for each call

**Comment by AThousandShips** [comment]
However it's a compatibility breaking change (as others depend on the current behaviour) so would have to be considered, also there isn't really any "correct" behaviour

Is there anything explicitly stated in the documentation? Otherwise it's really a matter of personal preference

**Comment by AThousandShips** [comment]
Also I cannot replicate your results, on MSVC the results are identical to GDScript

**Comment by ghsoares** [comment]
> Is there anything explicitly stated in the documentation? Otherwise it's really a matter of personal preference

Only reference of method chaining in the docs is for the Tween node, but there is no function that depends on internal state modification from previous calls that I could find.

> So this is compiler dependent in c++

I see, I've found a article addressing this issue before C++17: https://www.cppstories.com/2021/evaluation-order-cpp17/, but it seems that C++17 it became more strict the evaluation order to match the order I've expected, I don't know which version of C++ MSVC uses, but maybe this is the reason?

**Comment by AThousandShips** [comment]
> I don't know which version of C++ MSVC uses, but maybe this is the reason?

Nope this happens on C++20 as well

**Comment by AThousandShips** [comment]
My bad will test with latest versions, forgot the incorrect operator

**Comment by AThousandShips** [comment]
> but there is no function that depends on internal state modification from previous calls that I could find.

It's not about that, it's that users might depend on the *current* behavior, just like you argue people depend on it, so changing it *now* would break things for *them*, as the specific order isn't declared anywhere there's no "correct" order to violate here, right? Other than personal preference.

**Comment by ghsoares** [comment]
> It's not about that, it's that users might depend on the current behavior, just like you argue people depend on it, so changing it now would break things for them, as the specific order isn't declared anywhere there's no "correct" order to violate here, right? Other than personal preference.

I don't get it completely, could you explain in which case someone would depend on current behaviour?

**Comment by AThousandShips** [comment]
I mean, if they expect it to behave as it does right now, then changing it will break their code

If you want to be perfectly sure you won't be having these problems I'd suggest not chaining the calls, it's cleaner anyway

**Comment by AThousandShips** [comment]
Yes, I agree with all this, but that doesn't remove the fact that it still is a matter of breaking compatibility IMHO, it might not be considered so by others, but I'm just making it clear that this makes the situation more complicated in respect to when and if to change it, especially since it's not really some objective rule for this

> It's a matter of personal choice

I understand this, I'm just saying that if you want it to work in a specific way that's one way to accomplish that

**Comment by ghsoares** [comment]
I see, I would understand if it can't be implemented to evaluate the arguments only when calling the respective function, if it could it would be great to my library, but I'm okay with only being specified in the documentation the evaluation order, I only created this issue because I couldn't find any statement about this order so I interpreted it as a bug. Thanks for the clarification!

**Comment by AThousandShips** [comment]
No problem! I'm mostly assuming here but documenting it is a good idea, and discussing changing it as well, but chances are it'll take to either later 4.x or even 5.0
